### PR TITLE
fix!: do not clear selection in reset

### LINKED ISSUE #2674: CheckboxGroup.getDataProvider.refreshAll() clears selected values
The same issue described in #1674, #2593 (both triaged as `bug`) applies to `CheckBoxGroup` in Vaadin 22.0.4

https://github.com/vaadin/flow-components/blob/184cf30596ee41166e4d3493a2a288b9ccb3daad/vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java#L466-L468

### Description

Selected values are cleared after calling `getDataProvider().refreshAll()` 

### Expected outcome

[There should be some easy way of making a component reflect updates in the items (e.g. adding items or updating the contents of some item)](https://github.com/vaadin/flow-components/issues/1674#issuecomment-734810174)

### Actual outcome

Calling `getDataProvider().refreshAll()`  clears the value and one needs to set the old value after `refreshAll()` is called.

### Minimal reproducible example

```
@Route
public class MainView extends VerticalLayout {

    private RandomString rnd = new RandomString(3);

    public MainView(MultiSelect<CheckboxGroup<String>, String> listBox) {
        CheckboxGroup<String> group = new CheckboxGroup<>();

        List<String> items = new LinkedList<>(Arrays.asList("foo", "bar"));

        CheckboxGroupListDataView<String> dataView = group.setItems(items);

        Button button = new Button("Update items", e -> {           
            String newItem = rnd.nextString();
            dataView.addItem(newItem);
            dataView.refreshAll();  
        });

        add(group, button);

    }
}
```

### LINKED ISSUE #3131: CheckboxGroup :  calling setItemLabelGenerator(...) in a ValueChangeListener unchecks all checkboxes
### Description

In a ValueChangeListener of a CheckboxGroup the label of the items are refreshed by calling setItemLabelGenerator(...).
When clicking on one of the checkboxes, the labels are changed, but the checkboxes are all unchecked afterwards.

### Expected outcome

I would expect, that the checked state of the checkboxes remain unchanged, exclusing the one clicked.

### Minimal reproducible example

[CheckboxGroupView.java2.txt](https://github.com/vaadin/flow-components/files/8632285/CheckboxGroupView.java2.txt)

### Steps to reproduce

Click one of the unchecked checkboxes

### Environment

Vaadin version(s):  23.0.6
OS: Windows 10

### Browsers

Chrome, Firefox

### PR DESCRIPTION
**Pull request by ugur-vaadin**
## Description

Calling `DataProvider.refreshAll()` for `CheckboxGroup` used to clear the selected values.

This PR makes the `CheckboxGroup` retain the selection after `DataProvider.refreshAll()` is called and adds tests.

This is a breaking change.

Fixes #3131 
Fixes #2674

Similar to #2593

## Type of change

- [x] Bugfix
- [ ] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/contributing/overview
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

### CODE REVIEW

File: vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java [hunk 1]
```diff
@@ -693,6 +693,13 @@ private void reset() {
                 runBeforeClientResponse(sizeRequest);
             }
         }
+
+        Set<Object> currentItemIds = getCheckboxItems()
+                .map(CheckBoxItem::getItem).map(this::getItemId)
+                .collect(Collectors.toSet());
+        getValue().stream()
+                .filter(item -> !currentItemIds.contains(getItemId(item)))
+                .forEach(this::deselect);
```
Comment (tomivirkki) [inline vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java:702]: All the tests pass for me locally even if I remove this code

Comment (ugur-vaadin) [reply vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java]: Removed in the latest commit.

File: vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java [hunk 2]
```diff
@@ -358,11 +361,40 @@ public void setDataProvider(DataProvider<T, ?> dataProvider) {
...
                     }
                 });
     }
 
+    private void handleDataChange() {
+        Set<T> value = getValue();
+        if (value.isEmpty()) {
+            reset();
+            return;
+        }
```
Comment (tomivirkki) [inline vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java:374]: Is this condition tested somewhere?

File: vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java [hunk 3]
```diff
@@ -358,11 +361,40 @@ public void setDataProvider(DataProvider<T, ?> dataProvider) {
...
+        // Prevent probable loop caused by refreshing all data on value change
+        if (dataChangeHandlingInProgress) {
+            return;
+        }
```
Comment (tomivirkki) [inline vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java:378]: Can this loop condition be testes somehow?

File: vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java [hunk 4]
```diff
@@ -358,11 +361,40 @@ public void setDataProvider(DataProvider<T, ?> dataProvider) {
...
+        Set<T> obsoleteSelectedItems = value.stream()
+                .filter(item -> !currentItemIds.contains(getItemId(item)))
+                .collect(Collectors.toSet());
+        if (!obsoleteSelectedItems.isEmpty()) {
+            // Clear obsolete selected items
+            updateSelection(Collections.emptySet(), obsoleteSelectedItems);
+        } else {
+            // Force refresh selected items
+            clear();
+            setValue(value);
```
Comment (tomivirkki) [inline vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java:393]: Is this tested somewhere?

**Comment by ugur-vaadin** [comment]
Closing PR in favour of a new [one](https://github.com/vaadin/flow-components/pull/4503).

**Pull request closed without merging**